Avoid race conditions in MPTT update (parent_store)

rco-odoo opened the pull request:
Sometimes, the creation of a record in a hierarchical model (where `_parent_store=True`) introduces non-unique values for columns `parent_left`, `parent_right`.  This happens with transactions concurrently updating the MPTT fields.  This screws up the search on hierarchies.

The fix consists in constraining the combined values of the fields `parent_left` and `parent_right` to be unique.  We rely on the following invariant: in every pair `(parent_left, parent_right)`, either `parent_left` or `parent_right` is odd, and the other one is even.  The implementation creates two unique indexes: one for the odd values and the other for even ones.

The methods `create` and `write` have to be adapted to avoid violating the unique indexes in transient states. The current request is automatically retried when it fails because of a violation of this uniqueness constraint.

This fixes #14493.

rco-odoo commented:
@odony can you review this fix?
Note that I made it for version 10.0, but I can rebase it against version 8.0 if required.

KangOl commented on odoo/models.py:
```diff
@@ -3925,16 +3996,19 @@ def _create(self, vals):
+                # the nodes on the right are shifted in a way to avoid violating
+                # the uniqueness constraints on parent_left and parent_right:
+                # pos -> pos + M + 2 -> pos + 2
+                query = """ UPDATE {t} SET parent_left=parent_left+%(m)s+2 WHERE parent_left>=%(pleft)s;
```
For node creation, you can actually do it in 2 queries (shift + insert).
```sql
UPDATE {t} SET parent_left=parent_left+2, parent_right=parent_right+2 WHERE parent_left>=%s
UPDATE {t} SET parent_left=%s, parent_right=%s WHERE id=%s
```

KangOl replied:
You're right, the condition should be set on `parent_right`
```sql
UPDATE {t} SET parent_left=parent_left+2, parent_right=parent_right+2 WHERE parent_right>=%s
```

rco-odoo replied:
Sorry, you misunderstood me. The point is: in the example, the node A needs an update on its `parent_right` **only**. You should not update both `parent_left` and `parent_right` in this case.

rco-odoo replied:
You can actually do something a bit more tricky, like

    UPDATE {t}
    SET parent_left=parent_left+(CASE WHEN parent_left>=%s THEN 2 ELSE 0 END),
        parent_right=parent_right+(CASE WHEN parent_right>=%s THEN 2 ELSE 0 END)
    WHERE parent_left>=%s OR parent_right>=%s

With this query, you update either `parent_left`, or `parent_right`, or both, depending on whether they are to the right of the new node.

I have tried this query, but it did not work, but the bug was not clear. I may give it another try...

KangOl replied:
You may also reset the anchor node's parent_left after.
We are back to 3 queries as before.

rco-odoo replied:
Indeed, unique indexes are very restrictive for UPDATE queries:

    > CREATE TABLE stuff (id serial not null, z INTEGER);
    CREATE TABLE
    Time: 0.010s
    > CREATE UNIQUE INDEX stuff_z_index ON stuff (z);
    CREATE INDEX
    Time: 0.014s
    > INSERT INTO stuff (z) VALUES (1), (2), (3), (4);
    INSERT 0 4
    Time: 0.005s
    > SELECT * FROM stuff;
    +------+-----+
    |   id |   z |
    |------+-----|
    |    1 |   1 |
    |    2 |   2 |
    |    3 |   3 |
    |    4 |   4 |
    +------+-----+
    SELECT 4
    Time: 0.002s
    > UPDATE stuff SET z=z+2;
    duplicate key value violates unique constraint "stuff_z_index"
    DETAIL:  Key (z)=(3) already exists.

KangOl commented on odoo/models.py:
```diff
@@ -3731,18 +3741,76 @@ def _write(self, vals):
+                        pleft1 -= width
+                        # subtree nodes: pleft0 -> pleft0 + 2M -> pleft1
+                        # intermediate nodes: pos -> pos + M -> pos - width
+                        query = """ UPDATE {t}
```
Just like in `create`, you can shift nodes right to insertion point using only 1 query.

rco-odoo replied:
No you cannot. I had to fix this because it crashed: shifting `parent_left` has a high chance to introduce duplicates with non-shifted-yet `parent_right` values.

Consider the following example (`parent_left, parent_right` are given in parentheses):
- a node A (41, 44) with a single child B (42, 43).

Imagine that you have to shift that subtree 2 positions to the right in order to make some room for a single node. First, you update `parent_left`, and your nodes become:
- node A (43, 44) and node B (44, 45)

And BOOM, the unique index already rejects that because 44 appears twice!

rco-odoo commented on odoo/models.py:
```diff
@@ -2596,6 +2596,16 @@ def _parent_columns_exist(self):
+        # prevent race conditions during MPTT updates in record creation.
+        # /!\ Do not change the index names, as they are matched to retry
+        # transactions that failed because of them.
+        query = """ CREATE UNIQUE INDEX "{t}_parent_store_index0" ON "{t}"
```
1. This will make the point 2. more probable...
2. You have to explain me how to truncate the model name without taking the risk of a collision...

KangOl replied:
hash it.

Note that's already the case for other constraints...

KangOl replied:
Also, the index will only be set for new tables as the method wont be called if column `parent_left` already exists.

odony commented:
If possible we should keep the number of SQL queries very low, or at least make sure we never update twice the same record. These operations are typically expensive in terms of I/O + index write amplification, and we may be updating a tree with millions of records. In such a situation it's probably worth the trouble of having uglier/complicated queries if we can guarantee they will result in less I/O work on the disks.
That said, it's not worth redesigning the UNIQUE INDEXES, as that solution is the cleanest we have to guarantee the integrity.

rco-odoo commented:
I have reworked the branch, and changed the constraint that detects concurrent updates.
The unique index was not a correct solution (it allowed inconsistent results.)

We now use an EXCLUDE constraint to encode the following MPTT invariant: the intervals `(parent_left, parent_right)` of two nodes with the same parents should not overlap. This constraint makes concurrent updates fail as expected.

@odony @KangOl can you have a look at this new solution? Thanks!

odony commented on odoo/models.py:
```diff
@@ -2665,8 +2665,22 @@ def add(name, definition):
+
+        if self._parent_store:
+            # we need this extension
+            cr.execute('CREATE EXTENSION IF NOT EXISTS btree_gist')
```
For production use we recommend running with a non-super-user PG role, thus `CREATE EXTENSION` may not be allowed. We should guard against that and issue a warning in the log.
PG does not seem to raise an error if the user does not have `CREATE EXTENSION` privileges but the extension already exists and `IF NOT EXISTS` is specified, which makes it a little bit easier to handle.

rco-odoo replied:
Maybe we could move it to `base.sql`?

odony commented on odoo/models.py:
```diff
@@ -2665,8 +2665,22 @@ def add(name, definition):
+                         'COALESCE(parent_id, 0) WITH =,' \
+                         '(CASE WHEN parent_left<parent_right THEN int4range(parent_left, parent_right+1) ELSE NULL END) WITH &&' \
+                         ') INITIALLY DEFERRED'
+            sql_constraints.append(('parent_range', definition, ""))
```
I'll have to analyze that constraint definition further to be able to comment...
I wonder if the definition comparison system works well with this kind of definition, though? (i.e. it won't have to be re-created every time we call `auto_init`?)

rco-odoo replied:
The expression of the constraint would be simpler if `(parent_left, parent_right)` was stored as a single "range" column:

    EXCLUDE USING gist (COALESCE(parent_id, 0) WITH =, parent_range WITH &&)
        INITIALLY DEFERRED

The constraint is based on this "ideal" design ;-)

rco-odoo commented:
@mcb30 can you give a try to this branch?
I would like your confirmation that it works as expected before merging it.
Thanks a lot!

odony commented:
Closing this PR, as it was eventually superseded by #14694 and #22558.
